core/src/main/java/io/grpc/internal/SpiffeUtil2.java

```diff
@@ -0,0 +1,245 @@
+              + "Skipping certificate loading for trust domain '%s'.", KTY_PARAMETER_VALUE, kty,
+          trustDomainName));
+      return false;
+    }
+    String kid = JsonUtil.getString(jwkNode, "kid");
+    if (kid != null && !kid.equals("")) {
+      log.log(Level.SEVERE, String.format("'kid' parameter must not be set but value '%s' "
+              + "found. Skipping certificate loading for trust domain '%s'.", kid,
+          trustDomainName));
+      return false;
+    }
+    String use = JsonUtil.getString(jwkNode, "use");
+    if (use == null || !use.equals(USE_PARAMETER_VALUE)) {
+      log.log(Level.SEVERE, String.format("'use' parameter must be '%s' but '%s' found. "
+              + "Skipping certificate loading for trust domain '%s'.", USE_PARAMETER_VALUE, use,
+          trustDomainName));
+      return false;
+    }
+    return true;
+  }
+
+  private static List<X509Certificate> extractCert(List<Map<String, ?>> keysNode,
+      String trustDomainName) {
+    List<X509Certificate> result = new ArrayList<>();
+    for (Map<String, ?> keyNode : keysNode) {
+      if (!checkJwkEntry(keyNode, trustDomainName)) {
+        break;
+      }
+      String rawCert = JsonUtil.getString(keyNode, "x5c");
+      if (rawCert == null) {
+        break;
+      }
+      InputStream stream = new ByteArrayInputStream(rawCert.getBytes(StandardCharsets.UTF_8));
+      try {
+        Collection<? extends Certificate> certs = CertificateFactory.getInstance("X509")
+            .generateCertificates(stream);
+        if (certs.size() != 1) {
+          log.log(Level.SEVERE, String.format("Exactly 1 certificate is expected, but %s found for "
+                  + "domain %s.", certs.size(), trustDomainName));
+        } else {
+          result.add(certs.toArray(new X509Certificate[0])[0]);
+        }
+      } catch (CertificateException e) {
+        log.log(Level.SEVERE, String.format("Certificate for domain %s can't be parsed.",
+            trustDomainName), e);
+      }
+    }
+    return result;
+  }
+
+  // Will be merged with other PR
+  public static class SpiffeId {
+
+    private final String trustDomain;
+    private final String path;
+
+    private SpiffeId(String trustDomain, String path) {
+      this.trustDomain = trustDomain;
+      this.path = path;
+    }
+
+    public String getTrustDomain() {
+      return trustDomain;
+    }
+
+    public String getPath() {
+      return path;
+    }
+  }
+
+  /**
+   * Represents a Trust Bundle inspired by SPIFFE Bundle Format standard. Only trust domain's
+   * sequence numbers and x509 certificates are supported.
+   * @see <a href="https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Trust_Domain_and_Bundle.md#4-spiffe-bundle-format">Standard</a>
+   */
+  public static final class SpiffeBundle {
+
+    private final ImmutableMap<String, Long> sequenceNumbers;
+
+    private final ImmutableMap<String, ImmutableList<X509Certificate>> bundleMap;
```

Comment:
That's a really interesting question. JDK API uses [arrays](https://docs.oracle.com/javase/7/docs/api/javax/net/ssl/X509TrustManager.html), and gRPC Java uses lists - https://github.com/grpc/grpc-java/blob/927d21541dbc34be245f09c8180358a94ae5c031/xds/src/main/java/io/grpc/xds/internal/security/certprovider/CertificateProvider.java#L46
I suspect the issue with Sets is that X509Certificate itself is an abstract class, so developers can override equals/hashcode and it might lead to bugs.
So I'd just prefer to leave it a List similar with surrounding classes